File: src/main/java/emufog/graph/Node.java
/*
 * MIT License
 *
 * Copyright (c) 2018 emufog contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package emufog.graph;

import java.util.List;

/**
 * Represents a general node of  graph with the basic functionality.
 * Can connect to other nodes via edges and contains the geographically placement on a plain.
 */
public abstract class Node {

    final NodeAttributes attributes;

    /**
     * emulation config for this node
     */
    private EmulationSettings emulationSettings;

    Node(NodeAttributes attributes) {
        this.attributes = attributes;
        this.attributes.setNode(this);
    }

    /**
     * Returns the emulation config associated. If there are none returns {@code null}.
     *
     * @return emulation config or {@code null}
     */
    public EmulationSettings getEmulationNode() {
        return emulationSettings;
    }

    /**
     * Returns the unique identifier of the node.
     *
     * @return unique identifier
     */
    public int getID() {
        return attributes.id;
    }

    /**
     * Returns a list of all edges associated with the node.
     *
     * @return list of the node's edges
     */
    public List<Edge> getEdges() {
        return attributes.edges;
    }

    /**
     * Returns the name of the node.
     *
     * @return name of the node
     */
    public abstract String getName();

    public abstract NodeType getType();

    /**
     * Returns the edge degree of the node.
     *
     * @return number of edges associated with the node
     */
    public int getDegree() {
        return getEdges().size();
    }

    /**
     * Returns the autonomous system consisting the node.
     *
     * @return node's AS
     */
    public AS getAS() {
        return attributes.as;
    }

    /**
     * Returns identification if this node can be emulated with existing config.
     *
     * @return true if it can be emulated, false otherwise
     */
    public boolean hasEmulationSettings() {
        return emulationSettings != null;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Node)) {
            return false;
        }

        Node other = (Node) o;

        return attributes.equals(other.attributes);
    }

    @Override
    public int hashCode() {
        return attributes.hashCode();
    }

    @Override
    public String toString() {
        return getName();
    }

    void setEmulationSettings(EmulationSettings settings) {
        emulationSettings = settings;
    }
}

File: src/main/kotlin/emufog/graph/Node.kt
/*
 * MIT License
 *
 * Copyright (c) 2018 emufog contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package emufog.graph

/**
 * Represents a general node of the graph with the basic functionality. Can connect to other nodes via edges. Can also
 * hold an [EmulationNode] if that node should be emulated by a container.
 *
 * @property id unique identifier of the node
 * @property system the autonomous system this node belongs to
 * @property emulationNode emulation configuration of the node, can be not set
 * @property edges list of edges connected to this node
 * @property degree the edge degree of the node. Is based on the number of nodes this node is connected to via edges
 * @property type the type of this node
 */
abstract class Node internal constructor(
    val id: Int,
    val system: AS,
    edges: List<Edge>,
    emulationNode: EmulationNode?
) {

    private val edgesMutable: MutableList<Edge> = edges.toMutableList()

    val edges: List<Edge>
        get() = edgesMutable

    val degree: Int
        get() = edges.size

    open val emulationNode: EmulationNode?
        get() = emulationNodeMutable

    private var emulationNodeMutable: EmulationNode? = emulationNode

    abstract val type: NodeType

    /**
     * Returns whether this node is associated with an emulation configuration.
     */
    fun hasEmulationSettings(): Boolean = emulationNode != null

    internal fun setEmulationNode(emulationNode: EmulationNode?) {
        emulationNodeMutable = emulationNode
    }

    /**
     * Adds an [Edge] instance to the list of edges [edges].
     */
    internal fun addEdge(edge: Edge) {
        edgesMutable.add(edge)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is Node) {
            return false
        }

        return id == other.id
    }

    override fun hashCode(): Int = id

    override fun toString(): String = "Node: $id"
}
